### DISCUSSION TITLE
Display Image on NHD-4.3-480272FT-CTXL-T 4.3” EVE2 TFT Module (SPI)

### DISCUSSION THREAD

**Discussion by adityaseelam** [RudolphRiedel/FT800-FT813#115] [General]
Hello.
I'm new to this field. Ive an FPGA board with Kria K26 SOM and NHD-4.3-480272FT-CTXL-T 4.3” Screen. I just want to display image on that screen. Ive included all the header files(EVE.h,EVE_commands.c,EVE_config.c,EVE_cpp_target.cpp,EVE_cpp_wrapper.h,EVE_cpp_wrapper.cpp,EVE_supplemental.c,EVE_supplemental.h,EVE_target.c,EVE_target.h).Im running the application integration in Vitis IDE. Ive used  EVE Asset builder to convert my png to RGB565 format but I didnt compress it. Ive used BIN2C in EVE Asset builder to get the C file array. Im not going to print any logo so Should I just remove the logo array from tft_data.c? If so then what should my main.c file should be like? Do i need to add any other header files? Im sorry Im lost.

**Comment by RudolphRiedel** [comment]
Hi,

>Ive an FPGA board with Kria K26 SOM

This one?
https://www.amd.com/en/products/system-on-modules/kria/k26/k26c-commercial.html

What the heck? :-)
Consider me surprised, this is a first for me. :-)
That is so far out of scope for the EVE chips that I won't even ask why.

At this point the logo should be your least concern, or converting images.

So you need a target abstraction for this thing.
In the highly unlikely event that someone would give me something like this at work with the task to use it with
an EVE display, the first thing I would ask is on which of the several ARM cores the display thread is supposed to run.
Then the question is how to use GPIO pins and the SPI with this core.

I would try to start with a simple example program that already compiles.
I would check what compiler is used and which macros are set by the compiler to identifiy it.

Then I would add a new target file, name it perhaps EVE_target_Kria_K26.h, perhaps derived from EVE_target_template.h.
With the functions mostly empty like EVE_target_Tricore_Tasking.h unfortunately still is.
I would then add a folder to my project like this: [EmbeddedVideoEngine](https://github.com/RudolphRiedel/FT800-FT813/tree/5.x/examples/EVE_Test_SAMC21_EVE2-50G/EmbeddedVideoEngine)
And without tft.c or tft_data.c or anything actually calling the library, the first step is to make this compile.

Add a new target to EVE_target.h to include EVE_target_Kria_K26.h with likely at least including <stdint.h>.
Add a macro to your build options for the display-configuration you want, so likely EVE_NHD_43, or modify EVE_config.h to define this macro at the beginning.

Then figure out the GPIO, modify the target header accordingly.
Include EVE.h in your main.c and add calls to EVE_pdn_set(), EVE_pdn_clear(), EVE_cs_set() and EVE_cs_clear().
Make it compile, then check the designated pins with a logic-analyzer if this works,

Next modify spi_transmit() to work with your hardware and check if this works from your main.c:
EVE_cs_set()
spi_transmit(0x55)
spi_transmit(0x55)
EVE_cs_clear()

The SPI clock in this phase must be below 11MHz.

The next function that needs to work is spi_receive(), for this you could connect MISO and MOSI for a loopback, if you are not sure that the pins are configured correctly add a series resistor.
To check the result you could either print it over a serial connection or modiy the data in a loop:

uint8_t out = 0;
uint8_t in = 0xff;

loop
{
 in = spi_receive(out);
out = in + 1; // or be more creative
}

When the target header works, go further, try
EVE_cmdWrite(EVE_CLKSEL, 0x46U);
EVE_memWrite32(0x102030, 0x40506070);

Check the result with the logic-analyzer.

If that works, call EVE_init() and if this looks good, connect the display and use tft.c as is.
I recommend 56R series resistors for the SPI and GPIO lines.

Hmm, wow, good luck! :-)

**Comment by adityaseelam** [comment]
Hello,
Yes that's the one which I'm using  in my fpga board. Already my fpga board has designated pins which are compatible for this screen. I've tested these pins and I got the expected signals. I'll be honest with you, I've tried the libraries which are there here https://github.com/blazer82/FT81x_Arduino_Driver but apparently it only works for his image data when I try to compress my png to rgb565 and then to c array then I'm not able to get image data. So I wanted to use different libraries so I got here. I'm a beginner so I'll try to read and understand what you have sent to me.
Thanks.
Regards.,

**Comment by RudolphRrr** [comment]
Wait, what, there is an Arduino core for your FPGA multi ARM core AI image processing board? :-)
If that other library works, mine should be as well, at least for Arduino, my library has Arduino support, specialized for a variety of different Arduinos but also a generic "driver".

**Reply by adityaseelam** [reply]
Hello.,
Ive done it! I was able to display images using the different library the one which I've shared above(the repository) I was not able to do so before because I was using the jpg which is of 300*300[width*height]. I guess it might have been set to some limit so it doesnt take more than 300*300...but it should be able to take 480*272 size atmost..so maybe I need to do some changes in the loadimage function in my .c file. But anyways thank you.
P.S:The cmd_loadimage automatically compress the jpg data array into rgb565 data array so we need to only dump the jpg data in the array thats it, Dumb me was trying to dump the rgb565 data array in the code!

Regards.,

**Reply by RudolphRrr** [reply]
How did you use that linked library to load your image?
Did you modify it to support your board?

Converting the image with EAB has the benefit that you can decide which format you want to use and you can check the resulting size.
Well, for RGB565 the size is X * Y * 2.
And with FT81x / BT81x having 1MiB of RAM_G you could in theory put in a 800x600 image and scroll it around on the 480x272 panel, the only limit on EVEs side is the memory available.
Hmm, actually, what filesize has the .jpg that loads and how big is that .jpg that fails to load?

As for loading converted images, I recommend using the "ASSET COMPRESSOR" in EAB and then CMD_INFLATE.

**Reply by adityaseelam** [reply]
Hello,
Yes Ive changed the load image function according to my board.
The jpg which loads is of size 200X200 and The jpg which doesnt load is of size 400X300.
I failed to understand what does CMD_INFLATE do, what do you mean by converted images? (PNG->JPG is what converted is or am I missing something)?

Regards.,

**Comment by RudolphRrr** [comment]
I meant the size of the files in bytes, not in pixels.
400x300 in pixels should be fine though, that is 240000 bytes after loading with CMD_LOADIMAGE.

CMD_LOADIMAGE converts .jpg and .png to native formats.
And with PNG it is a bit on the slow side and also a bit picky about the format.

CMD_INFLATE only unzips binary data that was compressed before with zlib, so lossless compression.

And when you convert an image with the EVE Asset Builder the result is a raw image in the format you setup, ARG1555 or RGR565 or L8, or...
And these raw images can be compressed with the "ASSSET COMPRESSOR" in EVE Asset Builder and then used with CMD_INFLATE.

**Reply by adityaseelam** [reply]
Hello
Okay thank you for clearing my doubt@RudolphRiedel.

Regards,.
